qemu/tests/thin_provisioning_image.py

```diff
@@ -0,0 +1,165 @@
+        params["drive_format_local_disk"] = "scsi-hd"
+        params["drv_extra_params_local_disk"] = "discard=on"
+        object_params = params.object_params("local_disk")
+        object_params["image_name"] = "images/{}".format(params["image_name_local_disk"])
+        object_params["image_format"] = params["image_format_local_disk"]
+        object_params["image_size"] = params["image_size_local_disk"]
+        image = qemu_storage.QemuImg(object_params, base_dir, params["image_name_local_disk"])
+        image_file_name = image.create(object_params)[0]
+        params["image_name_local_disk"] = os.path.splitext(image_file_name)[0]
+
+    error.context("TEST STEP 2: Boot guest with image disk {:s}".format(image_file_name), logging.info)
+    env_process.preprocess_vm(test, params, env, vm_name)
+    vm = env.get_vm(vm_name)
+    vm.verify_alive()
+    timeout = float(params.get("login_timeout", 240))
+    session = vm.wait_for_login(timeout=timeout)
+
+    guest_device_name = thin_provisioning.get_guest_discard_disk(session)
+    format_all_disk_cmd = params["format_all_disk_cmd"]
+    format_all_disk_cmd = format_all_disk_cmd.replace("DISK1", guest_device_name)
+    error.context("TEST STEP 3: Format disk '{:s}' in guest".format(guest_device_name), logging.info)
+    session.cmd(format_all_disk_cmd)
+    block_num_after_format = count_block_num(image_file_name)
+    error.context("block_num_after_format: {:s}".format(block_num_after_format), logging.info)
+
+    mount_disk_cmd = params["mount_disk_cmd"]
+    error.context("TEST STEP 4: Mount disk to {:s} in guest.".format(mount_disk_cmd), logging.info)
+    mount_disk_cmd = mount_disk_cmd.replace("DISK1", guest_device_name)
+    session.cmd(mount_disk_cmd)
+
+    error.context("TEST STEP 5: Create a file in the mount directory.", logging.info)
+    ddfile = "{:s}/file".format(params.get("mount_point"))
+    rewrite_disk_cmd = "dd if=/dev/zero of={:s} bs=1M count=2000".format(ddfile)
+    try:
+        session.cmd(rewrite_disk_cmd, timeout=timeout)
+    except:
+        raise error.TestError("Create file in guest failed.")
+    block_num_after_dd = count_block_num(image_file_name)
+    error.context("block_num_after_dd: {:s}".format(block_num_after_dd), logging.info)
+
+    session.cmd("rm -rf %s" % ddfile)
+    time.sleep(5)
+    error.context("TEST STEP 6: Execute fstrim in guest", logging.info)
+    fstrim_cmd = params["fstrim_cmd"]
+    session.cmd(fstrim_cmd, timeout=timeout)
+    block_num_after_trim = count_block_num(image_file_name)
+    error.context("block_num_after_trim: {:s}".format(block_num_after_trim), logging.info)
+    if params.get("image_format_local_disk") == "qcow2" or params.get("image_format_gluster_disk") == "qcow2":
+        if params.get("guest_filesystem") == "ext4":
+            if block_num_after_trim != "277520":
+                raise exceptions.TestFail("Expect 277520, but we got {:s}".format(block_num_after_trim))
```

Comment:
14:04:47 ERROR| TestFail: Expect 277520, but we got 276880
14:04:47 ERROR|
14:04:47 ERROR| FAIL 6-smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.thin_provisioning_image.ext4.local_image.local_disk_qcow2 -> TestFail: Expect 277520, but we got 276880